Issue 3652: Segment store TLS reload upon certificate refresh

ravisharda opened the pull request:
**Change log description**  
Reload TLS when segment store TLS material is refreshed. 

**Purpose of the change**  
Resolves #3652 

**What the code does**  
1. Introduces a new configuration flag in Segment Store that determines whether to reload TLS when the TLS certificate changes. 
2. The flag is off by default. If the flag is on: 
    1. Watches for change in Segment Store TLS certificate. 
        We use an event-based WatchService to watch for file modification events. That mechanism doesn't work when the file being watched is a symbolic link. Certificate management tools such as Cert Manager distribute cert as a symbolic link, so in such cases, we use a polling-based file modification monitor.
    2. Upon detecting that the certificate has changed, we replace the shared SSL context with a new SSL Context. Existing connections that were set up using the old context continue to function, while new connections use an SSL Context built from the new certificate and key.
3. The above is implemented both for distributed mode and standalone mode cluster. 
    It is supported in standalone mode to make it easier to debug and test the feature. The flag is intentionally left out of the `standalone-config.properties` file, but is a hidden flag and may be added by developers if needed. 

**How to verify it**  
All unit tests should pass. 

In addition to unit tests, I've executed the following types of tests to verify the functionality:

* Standalone mode cluster with client application writing to the cluster continuously. Replaced the certificate midway and check the client continues to function. Also, run a new client application.
* Docker-Compose cluster with the same scenario as above.
* Kubernetes cluster with Cert Manager distributing certificates to Segment Store, with the same scenario as above.

In (3), the certificate path was a symbolic link. In (1) and (2), it was a direct path (no symlink).

tkaitchuck commented on common/src/main/java/io/pravega/common/io/FileModificationWatcher.java:
```diff
@@ -0,0 +1,126 @@
+     */
+    private final Consumer<WatchEvent<?>> callback;
+
+    private UncaughtExceptionHandler uncaughtExceptionalHandler = (t, e) -> handleException(t.getName(), e);
```
Final

ravisharda replied:
Done.

andreipaduroiu commented on common/src/main/java/io/pravega/common/io/FileModificationWatcher.java:
```diff
@@ -0,0 +1,126 @@
+
+        this.pathOfFileToWatch = Paths.get(fileToWatch);
+
+        if (!this.pathOfFileToWatch.toFile().exists()) {
```
FileNotFoundException

ravisharda replied:
Sure, thanks for noticing it. I've changed it to FileNotFoundException in the latest commit.

andreipaduroiu commented on common/src/main/java/io/pravega/common/io/FileModificationWatcher.java:
```diff
@@ -0,0 +1,126 @@
+                            .stream()
+                            .filter(event -> // we only care about changes to the specified file.
+                                    event.context().toString().contains(fileName))
+                            .forEach(event ->  callback.accept(event)); // invoke the specified callback
```
too many consecutive spaces

ravisharda replied:
Fixed.

andreipaduroiu commented on common/src/main/java/io/pravega/common/io/FileModificationWatcher.java:
```diff
@@ -0,0 +1,126 @@
+            log.debug("Done setting up watch for modify entries.");
+
+            while (true) {
+                watchKey = watchService.take();
```
I assume this call will block the thread?

I'm trying to understand why you need a dedicated thread for this.

ravisharda replied:
The reason I've used a thread for this class is so that the caller can use it asynchronously and continue to perform other tasks and not get blocked.

andreipaduroiu commented on common/src/main/java/io/pravega/common/io/FileModificationWatcher.java:
```diff
@@ -0,0 +1,126 @@
+ *
+ */
+@Slf4j
+public class FileModificationWatcher extends Thread {
```
If you really **must** use up an entire thread just for this watcher, please give it a proper name. That way we can filter out by its name in the logs.


Please consider what happens when you have multiple of these running - make sure you don't clash names.

ravisharda replied:
Sure, I've added a thread name in the latest commit.

andreipaduroiu commented on common/src/main/java/io/pravega/common/io/FileModificationWatcher.java:
```diff
@@ -0,0 +1,126 @@
+                boolean isKeyValid = watchKey.reset();
+                log.debug("Done resetting watch key, so that it can receive further event notifications.");
+                if (!isKeyValid) {
+                    log.info("No longer watching file [{}]", this.pathOfFileToWatch);
```
So how do you exit this while loop? I would have expected this to break from it, otherwise it would keep on looping forever.

ravisharda replied:
Sure, I've modified the code to exit the loop at this point.

andreipaduroiu commented on common/src/main/java/io/pravega/common/io/FileModificationWatcher.java:
```diff
@@ -0,0 +1,126 @@
+
+            while (true) {
+                watchKey = watchService.take();
+                log.debug("Done setting up watch key.");
```
Add file name, here and everywhere else you log.

ravisharda replied:
Done.

andreipaduroiu commented on common/src/main/java/io/pravega/common/io/FileModificationWatcher.java:
```diff
@@ -0,0 +1,126 @@
+                }
+            }
+        } catch (IOException | InterruptedException e) {
+            log.warn(e.getMessage(), e);
```
Double-logging the exception.

Why don't you use `handleException`?

ravisharda replied:
Fixed.

andreipaduroiu commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/Channels.java:
```diff
@@ -0,0 +1,81 @@
+@Slf4j
+public class Channels {
+
+    private Lock sequential = new ReentrantLock();
```
final

ravisharda replied:
Fixed.

andreipaduroiu commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/Channels.java:
```diff
@@ -0,0 +1,81 @@
+    private Lock sequential = new ReentrantLock();
+
+    // See info about ChannelGroups here: https://netty.io/4.1/api/io/netty/channel/group/ChannelGroup.html
+    private ChannelGroup channels =
```
final

ravisharda replied:
This variable is not intended to be `final`. It is reassigned to a new instance in `flushStopAndRefresh(...)` method.

andreipaduroiu commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/Channels.java:
```diff
@@ -0,0 +1,81 @@
+        try {
+            future.await();
+        } catch (InterruptedException e) {
+            log.warn(e.getMessage(), e);
```
Double-logging

ravisharda replied:
Fixed.

andreipaduroiu commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/Channels.java:
```diff
@@ -0,0 +1,81 @@
+            log.info("Connection cancelled by user.");
+        } else if (!future.isSuccess()) {
+            Throwable e = future.cause();
+            log.warn(e.getMessage(), e);
```
and here

ravisharda replied:
Fixed.

andreipaduroiu commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/Channels.java:
```diff
@@ -0,0 +1,81 @@
+            log.warn(e.getMessage(), e);
+            // Ignore. Intentionally not rethrowing the exception, so that the system continues to function.
+        } else {
+            log.info("Done closing the channel group, which implies that all channels in the group were "
```
Please rephrase these log messages in a different way. They all start with "Done with", which will make it hard to search for anything in the logs. 

Also make them succinct, don't include an explanation with each one of them. You can do that in a code comment.

ravisharda replied:
Sure, done.

andreipaduroiu commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaConnectionListener.java:
```diff
@@ -41,53 +41,65 @@
      * Creates a new instance of the PravegaConnectionListener class listening on localhost with no StatsRecorder.
      *
-     * @param ssl                Whether to use SSL.
+     * @param enableTls       Whether to use SSL.
```
Indentation

ravisharda replied:
Fixed.

andreipaduroiu commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaConnectionListener.java:
```diff
@@ -187,5 +239,8 @@ public void close() {
         // Shut down all event loops to terminate all threads.
         bossGroup.shutdownGracefully();
         workerGroup.shutdownGracefully();
+        if (!executor.isShutdown()) {
```
This is the wrong way to shut down an executor. See `ExecutorServiceHelpers.shutdown`

ravisharda replied:
I've modified the code to avoid using the executor, as per another suggestion below.

andreipaduroiu commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaConnectionListener.java:
```diff
@@ -173,6 +211,20 @@ public void initChannel(SocketChannel ch) throws Exception {
+                    new TLSConfigChangeEventConsumer(channels));
+            tlsCertFileChangeWatcherTask.setDaemon(true);
+
+            this.executor.submit(tlsCertFileChangeWatcherTask);
```
So you have a `FileModificationWatcher` which extends `Thread`, which you then submit to an `ExecutorService` which has its own threads. That's one thread too many. 

I suggest you remove the executor from this class and simply execute your one-off thread. But if you do that, you need to keep a reference to it so that when you do close this class, you interrupt that thread.

ravisharda replied:
Yes, I agree. I've changed it in the latest commit.

andreipaduroiu commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/TLSConfigChangeEventConsumer.java:
```diff
@@ -0,0 +1,42 @@
+
+@RequiredArgsConstructor
+@Slf4j
+public class TLSConfigChangeEventConsumer implements Consumer<WatchEvent<?>> {
```
Javadoc

ravisharda replied:
Done.

andreipaduroiu commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/TLSConfigChangeEventConsumer.java:
```diff
@@ -0,0 +1,42 @@
+public class TLSConfigChangeEventConsumer implements Consumer<WatchEvent<?>> {
+
+    /**
+     * A counter representing the number of times this object has been asked to
```
This comment is wrong. You refer to `this object` but this field is static. I don't think you should make it static

ravisharda replied:
Yeah, initially an instance of this class was supposed to be created for each Netty pipeline. So, the counter needed to be static. Now, we register the watcher just once, so there is no need to make this static. I've fixed this in the latest commit.

andreipaduroiu commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/TLSConfigChangeEventConsumer.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Override
+    public void accept(WatchEvent<?> watchEvent) {
+        log.debug("Invoked for [{}]", watchEvent.context().toString());
```
Do not call `toString`. `log.debug` will automatically call it if debug-logging is enabled. Otherwise you will invoke that with every call.

ravisharda replied:
Fixed.

andreipaduroiu commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/TLSConfigChangeEventConsumer.java:
```diff
@@ -0,0 +1,42 @@
+    private @NonNull final Channels channels;
+
+    @Override
+    public void accept(WatchEvent<?> watchEvent) {
```
Where do you use this argument?

ravisharda replied:
The `FileModificationWatcher` object to which we pass an object of `TLSConfigChangeEventConsumer` as a callback, provides the `WatchEvent` as context to the callback. Currently `TLSConfigChangeEventConsumer` uses the `WatchEvent` only for logging the name of the file for which the modification was detected. 

The newly introduced `FileModificationWatcher` is a generic class that can be useful in other places too, and I wanted to keep its interface useful for general use. So, I'm passing the WatchEvent as a callback. 

What do you say?

andreipaduroiu commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/TLSHelper.java:
```diff
@@ -0,0 +1,47 @@
+import javax.net.ssl.SSLException;
+
+@Slf4j
+public class TLSHelper {
```
Javadoc

ravisharda replied:
Done.

ravisharda commented:
> 1. Where are the unit tests for the new classes?

One of the reasons I left the PR in draft status was that tests weren't added yet. I've now added some unit tests. I shall be adding more soon. 

> 2. I recommend you move them out of `segmentstore:server:host` and into some other project. The `host` project should be responsible only with putting everything together and not have any application logic. I consider security-related stuff as part of that category.

Sure, that sounds good to me. I've moved the classes into package `io.pravega.segmentstore.server.security` in `segmentstore:server` sub-project. Does that look OK?

EronWright commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaConnectionListener.java:
```diff
@@ -144,26 +155,48 @@ public void startListening() {
+                         // Creating an SSL Context for each init ensures that the latest cert and key files are used for
+                         // any new channels.
+                         sslContext = TLSHelper.newServerSslContext(pathToTlsCertFile, pathToTlsKeyFile);
+                         log.debug("Created a new SSL context.");
```
It is too expensive to create a new `SslContext` on each request.   The best you can do is, use the watch events to update the cached context, and use the cached context whenever creating a new request.

ravisharda replied:
I ran a few tests creating an `io.netty.handler.ssl.SslContext.SslContext` using the `io.netty.handler.ssl.SslContextBuilder.forServer(...)` method used in Pravega, and found that it wasn't expensive - at least in terms of the time it takes to create each instance.  One such test was creating `SslContext` 200 times in a loop. It took ~1.3 second the first time, and 96 ms. on average for the rest of the 199 times. 

The call to create a new SslContext happens once per channel. Also, we don't create channels on a channel-per-request. It is shared across multiple requests. They get closed either at the client application's behest or if an exception/error causes the server to close the channel.  

I suspect that updating the cached context can be risky, especially for in-flight requests. And, there might be different consequences depending on the stage the request is at. I find it easier to articulate and reason about the consequences of TLS certificate change with the simplistic model of handling the certificate modification that is there in the PR code today:

* If TLS reload is enabled:
   * If the certificate changes, existing channels will be closed by the server, resulting in the client library code encountering connection drops. Upon connection drops, the library code automatically retries connecting to the server, and this time a new SslContext built using the latest certificate is used. 
  * Regardless of whether the certificate remains the same or modifies, all new channels are created using a SslContext built from the latest certificate. 

   As a result, a new certificate is immediately applied to all existing and new connections. The need to refresh existing channels stems from the need to handle expired certificates too.

* If TLS reload is disabled:
   * SslContext is created once and the same cached instance is used for every channel.  

What do you say?

EronWright replied:
This is not a good take.  You cannot create an `SslContext` per channel.  Channels are hardly as reusable as you suggest.   A simple `ReentrantReadWriteLock` (fair mode) will suffice to protect the cached SSL context; obtain the (shared) read lock whenever a channel is created, and obtain the write lock on the watch thread (after creating the replacement context but before updating the cache).

Do not close existing channels.

ravisharda replied:
Upon a deeper look into the pros and cons of the two approaches, I think the approach you have suggested is simpler. I have modified the code and pushed the changes.

EronWright commented on segmentstore/server/src/main/java/io/pravega/segmentstore/server/security/TLSConfigChangeEventConsumer.java:
```diff
@@ -0,0 +1,45 @@
+
+    private void handleTlsConfigChange() {
+        log.info("Current reload count = {}", numOfConfigChangesSinceStart.incrementAndGet());
+        channels.clear();
```
I don't see the value in closing existing connections.  That said, an idle timeout on the connection would be handy to gracefully cleanup old connections.

ravisharda replied:
Closing existing channels will result in the client library code encountering connection drops. Upon connection drops, the library code automatically retries connecting to the server, and this time a new SslContext built using the latest certificate is used.

Channels can live for long. If they continue to use SslContext built from old certificates that have expired or are revoked, there shall likely be knotty issues to deal with w.r.t. error handling.

EronWright replied:
@ravisharda your argument is not strong without an example of a 'knotty issue'.    I see more potential for knotty issues in closing the connection.

Once TLS negotiation is complete and the channel is active for data transfer, there is no practical reason to close the channel.

gaddamas replied:
Agree with Eron's comment. Consider it synonymous with filesystem calls, once you have file handle open for read/write it can be used even if the file has been renamed or deleted. File will actually be deleted only after the last file handle is closed.
Typically with expired certs problem is authentication, not the encryption they provide.
Since the session is already authenticated during the handshake before the cert was expired it should be good to use it as long as the connection is open.

ravisharda replied:
I've modified the implementation, as mentioned in a previous comment.

ravisharda commented:
> The `SslContext` must be cached.

Please see this comment: https://github.com/pravega/pravega/pull/3966#discussion_r299310404. Please take a look and let me know what you think about it.

shrids commented on segmentstore/server/src/main/java/io/pravega/segmentstore/server/security/Channels.java:
```diff
@@ -0,0 +1,131 @@
+     *
+     */
+    public void clear() {
+        if (!this.channels.isEmpty()) {
```
This should be under the lock too?

ravisharda replied:
I removed this class now, as it is irrelevant given the change in strategy in favor of `SslContext` reload.

shrids commented on segmentstore/server/src/main/java/io/pravega/segmentstore/server/security/Channels.java:
```diff
@@ -0,0 +1,131 @@
+    private final Lock sequential = new ReentrantLock();
+
+    // See info about ChannelGroups here: https://netty.io/4.1/api/io/netty/channel/group/ChannelGroup.html
+    private ChannelGroup channels =
```
@GuardedBy

ravisharda replied:
I have removed this class now, as it is irrelevant given the change in strategy in favor of `SslContext` reload.

shrids commented on segmentstore/server/src/main/java/io/pravega/segmentstore/server/security/Channels.java:
```diff
@@ -0,0 +1,131 @@
+        log.debug("Triggered flush of channel group {}", channelGroup);
+        ChannelGroupFuture future = channelGroup.close();
+        try {
+            future.await();
```
Should we wait until the future completes?

ravisharda replied:
This class is not there anymore, as it is now irrelevant given the change in strategy in favor of `SslContext` reload.

ravisharda commented:
The last commit `1bdb461` changes the strategy for TLS reload to reloading the SSL Context. This specific change is not tested as of now. I've committed it so that reviewers can take an early look. This PR should not be merged until I confirm that it is tested.

EronWright commented on segmentstore/server/src/main/java/io/pravega/segmentstore/server/security/TLSConfigChangeEventConsumer.java:
```diff
@@ -0,0 +1,50 @@
+
+    private void handleTlsConfigChange() {
+        log.info("Current reload count = {}", numOfConfigChangesSinceStart.incrementAndGet());
+        sslContext.set(TLSHelper.newServerSslContext(pathToCertificateFile, pathToKeyFile));
```
This looks good, but please add some robustness here.   I believe that a watch event may come while the updated file is still being written, leading to an IOException (i.e. a race).   Please catch IOException here and schedule a retry.

ravisharda replied:
According to the [Java Doc](https://docs.oracle.com/javase/8/docs/api/java/nio/file/StandardWatchEventKinds.html#ENTRY_MODIFY): 
* ENTRY_MODIFY represents "Directory entry modified"
* ENTRY_CREATE represents "Directory entry created."

The `WatchDir` [example](https://docs.oracle.com/javase/tutorial/essential/io/examples/WatchDir.java) from Oracle also doesn't do anything special in this respect.

My understanding is that these events are raised only after the corresponding operation is done. For the `ENTRY_MODIFY` event, multiple events might be raised, usually once for content and the second time for metadata. I've added a sleep-based deduplication logic in  `FileModificationWatcher.java` to compensate for it. 

That being said, your comment prompted me to add some additional checks in `TLSHelper.newServerSslContext(...)` to ensure that the files exist and are readable.

ravisharda commented on common/src/main/java/io/pravega/common/io/FileModificationWatcher.java:
```diff
@@ -0,0 +1,190 @@
+            log.debug("Directory being watched is {}", directoryPath);
+
+            assert directoryPath != null;
+            directoryPath.register(watchService, StandardWatchEventKinds.ENTRY_MODIFY);
```
Sure, that makes sense. I've changed the code to look for `ENTRY_CREATE` event too.

ravisharda commented on common/src/main/java/io/pravega/common/io/FileModificationWatcher.java:
```diff
@@ -0,0 +1,190 @@
+                            .filter( // we only care about changes to the specified file.
+                                    event -> event.context().toString().contains(fileName))
+                            .forEach( // invoke the specified callback
+                                    filteredEvent -> callback.accept(filteredEvent));
```
Yes, we only care about one of the multiple events that may be raised at this point (even though the chances of multiple events being raised are small due to the deduplication logic above). I've changed the code accordingly.

tkaitchuck commented on common/src/main/java/io/pravega/common/io/FileModificationWatcher.java:
```diff
@@ -0,0 +1,195 @@
+                // Looks odd, right? Using the logic or de-duplicating file change events, as suggested by some here:
+                // https://stackoverflow.com/questions/16777869/java-7-watchservice-ignoring-multiple-occurrences-of-
+                // the-same-event
+                Thread.sleep(200);
```
Rather than having a thread per-file where each thread blocks in a loop. It would be better to have a pool where each task is run with a fixed delay. Then there can be fewer threads than files and while one is blocking another file can be checked.

See https://guava.dev/releases/snapshot/api/docs/com/google/common/util/concurrent/AbstractScheduledService.html
which provides an easy API to do this.

EronWright replied:
A task pool might also make it easier to reschedule the reload in the edge case of an IOException when reading the cert file due to a race.   I honestly think that file change notification may arrive while the file is still open for write.

ravisharda replied:
Currently, the `FileModificationWatcher` is used to monitor only a single file - TLS certificate file for Segment Store. Granted, it is a generic class that can find a use for other things as well, and that the change you are suggesting would be very useful. But, I think it might be prudent to enhance it further if and when we really need it. 

Would you agree?

tkaitchuck replied:
*If* we're sure there is only ever going to be one of these in a given process. A dedicated thread and a sleep is probably fine. Are we sure of that?

ravisharda replied:
As of now, it is going to be monitoring a single file, so there is going to be one of it in a given process. Things might change in the future, and we might have to enhance it then.

tkaitchuck commented on common/src/main/java/io/pravega/common/io/FileModificationWatcher.java:
```diff
@@ -0,0 +1,195 @@
+                }
+            }
+        } catch (InterruptedException e) {
+            log.info("Thread {}, watching for modifications in file {}, interrupted with cause {}",
```
This is not an appropriate way to handle an interrupted exception.
See `Exceptions.handleInterupted`

ravisharda replied:
I've modified this part of the code a bit. Please take a look again and see if it looks OK.

shrids commented on common/src/main/java/io/pravega/common/io/FileModificationWatcher.java:
```diff
@@ -0,0 +1,192 @@
+
+                // Looks odd, right? Using the logic or de-duplicating file change events, as suggested by some here:
+                // https://stackoverflow.com/questions/16777869/java-7-watchservice-ignoring-multiple-occurrences-of-
+                // the-same-event
```
`io.pravega.common.Exceptions#handleInterrupted`  can be used.

ravisharda replied:
I wanted to log a message when the condition occurs for debugging purposes, so have handled the exception using a catch block.

shrids commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaConnectionListener.java:
```diff
@@ -173,6 +192,23 @@ public void initChannel(SocketChannel ch) throws Exception {
+                log.info("Done setting up TLS reload, which in turn will be based on changes in file: {}",
+                        this.pathToTlsCertFile);
+            } catch (FileNotFoundException e) {
+                log.warn("Failed to setup a watcher on the file {}", this.pathToTlsCertFile, e);
```
we should change it to ERROR level since this would prevent the SSS service from starting.

ravisharda replied:
Fixed.

shrids commented on segmentstore/server/src/main/java/io/pravega/segmentstore/server/security/TLSConfigChangeEventConsumer.java:
```diff
@@ -0,0 +1,61 @@
+        if (watchEvent != null) {
+            log.info("Invoked for [{}]", watchEvent.context());
+        } else {
+            log.warn("Invoked for null watchEvent");
```
Can this be null?

ravisharda replied:
I've added this null check as a defensive check. It also ensures that the contract is explicit.

ravisharda commented:
> The last commit `1bdb461` changes the strategy for TLS reload to reloading the SSL Context. This specific change is not tested as of now. I've committed it so that reviewers can take an early look. Also, I haven't removed the code that would become irrelevant with this change: I'll do that after the new approach is found to be working effectively.
> 
> This PR should not be merged until I confirm that it is tested.

I've now tested this PR with both with:

* Standalone mode cluster where I updated the TLS cert manually 
* Distributed mode Kubernetes cluster with JetStack Cert Manager renewing the certificate automatically

ravisharda commented:
> Please bring up test coverage. (Only unit tests count towards the coverage report.)

@tkaitchuck thanks. I've added more unit tests, as you had suggested.

ravisharda commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java:
```diff
@@ -0,0 +1,206 @@
+ * Watches for modifications to the specified file and performs the specified action (in the form of a callback) upon
+ * modification detection.
+ *
+ * This implementation uses NIO and event-based Java WatchService to watch for any modifications to the specified file..
```
Fixed. Thanks for spotting it. :)

ravisharda commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java:
```diff
@@ -0,0 +1,206 @@
+                log.info("Retrieved and removed watch key for watching file at path: {}", this.pathOfFileToWatch);
+
+                // Looks odd, right? Using the logic or de-duplicating file change events, as suggested by some here:
+                // https://stackoverflow.com/questions/16777869/java-7-watchservice-ignoring-multiple-occurrences-of-
```
Sure, I've modified the comment now. Hope it makes a lot more sense now.

ravisharda commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java:
```diff
@@ -0,0 +1,206 @@
+                // Looks odd, right? Using the logic or de-duplicating file change events, as suggested by some here:
+                // https://stackoverflow.com/questions/16777869/java-7-watchservice-ignoring-multiple-occurrences-of-
+                // the-same-event
+                Thread.sleep(200);
```
Each file modification/create usually results in the WatcherService reporting the WatchEvent twice, as the file is updated twice: once for the content and once for the file modification time. These events occur in quick succession. We wait for 200 ms. here so that the events get
de-duplicated - in other words only single event is processed.

If the two events are not raised within this duration, the callback will be invoked twice, which we assume is not a problem for applications of this object. In its current use for reloading the SSL Context, that would mean reloading the context twice, which doesn't break anything; It's just that the SSL context is reloaded twice unnecessarily - reloading it once is sufficient.

ravisharda commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationPollingMonitor.java:
```diff
@@ -0,0 +1,138 @@
+            try {
+                monitor.stop(5 * 1000);
+            } catch (Exception e) {
+                log.warn("Failed in closing the monitor", e);
```
Fixed.

andreipaduroiu commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java:
```diff
@@ -0,0 +1,225 @@
+    }
+
+    @VisibleForTesting
+    String nameOfFileToWatch() {
```
`getWatchedFileName`

andreipaduroiu replied:
I would also suggest changing the names of the private fields as well, to make them easier to read.

ravisharda replied:
Sure, that makes sense now that you say it. I've modified the names.

andreipaduroiu commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java:
```diff
@@ -0,0 +1,225 @@
+        if (fileName != null) {
+            return fileName.toString();
+        } else {
+            throw new RuntimeException("File name is null");
```
How can this possibly be null?
And if so, why throw the exception here? 
Furthermore, why throw `RuntimeException` when it would have thrown a `NullPointerException` anyway which would have been more informative.

ravisharda replied:
Spotbugs complains with a `NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE` (Possible null pointer dereference) error if I don't check for null here. I've modified the code to now throw `NullPointerException` now.

andreipaduroiu commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java:
```diff
@@ -0,0 +1,225 @@
+    }
+
+    @VisibleForTesting
+    Path directoryOfFileToWatch() {
```
please change name here too.

ravisharda replied:
Done.

andreipaduroiu commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java:
```diff
@@ -0,0 +1,225 @@
+                            .stream()
+                            .filter( // we only care about changes to the specified file.
+                                    event -> event.context().toString().contains(fileName))
+                            .findAny(); // we only care to know whether the
```
Incomplete comment.

ravisharda replied:
:) I wonder how I missed that. I've fixed it now.

andreipaduroiu commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java:
```diff
@@ -0,0 +1,225 @@
+                    }
+
+                    boolean isKeyValid = watchKey.reset();
+                    log.debug("Done resetting watch key, so that it can receive further event notifications.");
```
I asked you in a previous review of this PR to make your logs less verbose. There is no need to include all these explanations in here. Just indicate at which point in your processing you are at (the bare minimum).

ravisharda replied:
I've modified it to make the log less verbose as suggested.

andreipaduroiu commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationMonitor.java:
```diff
@@ -0,0 +1,51 @@
+     * A convenience method for implementations.
+     *
+     * @param filePath the path of the file to be monitored
+     * @param callback the callback to be invoked when a file modification is detected
```
1. Why is this argument here? it's not used. If all you need to do is check for nulls, then annotate with `@NonNull` wherever you are accepting it.

2. Move this method in the implementation. There is no need for it to be exposed to the outside as it doesn't provide any meaningful use.

ravisharda replied:
Yes, that makes sense. It was looking awkward to me too, but I somehow missed modifying it.

andreipaduroiu commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java:
```diff
@@ -0,0 +1,225 @@
+
+    private final UncaughtExceptionHandler uncaughtExceptionalHandler = (t, e) -> logException(e);
+
+    private boolean loopContinuously;
```
`final`

ravisharda replied:
Done.

andreipaduroiu commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationPollingMonitor.java:
```diff
@@ -0,0 +1,138 @@
+     */
+    private final Consumer<File> callback;
+
+    private FileAlterationMonitor monitor;
```
final

ravisharda replied:
Done.

andreipaduroiu commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationPollingMonitor.java:
```diff
@@ -0,0 +1,138 @@
+
+    private FileAlterationMonitor monitor;
+
+    private int pollingInterval;
```
final

ravisharda replied:
Done.

andreipaduroiu commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationPollingMonitor.java:
```diff
@@ -0,0 +1,138 @@
+     * @param callback    the callback to invoke when a modification to the {@code fileToWatch} is detected
+     *
+     * @throws NullPointerException if either {@code fileToWatch} or {@code callback} is null
+     * @throws InvalidPathException if {@code fileToWatch} is invalid
```
Where is this being thrown from?

ravisharda replied:
`Paths.get(..)` may throw `InvalidPathException`, as described [here](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Paths.html#get-java.lang.String-java.lang.String...-).

andreipaduroiu commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationPollingMonitor.java:
```diff
@@ -0,0 +1,138 @@
+    @Override
+    public void startMonitoring() {
+        Path fileName = pathOfFileToWatch.getFileName();
+        log.debug("fileName obtained from pathOfFileToWatch is [{}]", fileName);
```
Proper casing in your log messages

ravisharda replied:
Fixed.

andreipaduroiu commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationPollingMonitor.java:
```diff
@@ -0,0 +1,138 @@
+    public void startMonitoring() {
+        Path fileName = pathOfFileToWatch.getFileName();
+        log.debug("fileName obtained from pathOfFileToWatch is [{}]", fileName);
+        if (fileName == null) {
```
How can this be?

ravisharda replied:
The API docs of `Path` says `getFileName()` can return null.  This null check is a defensive check.

andreipaduroiu commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationPollingMonitor.java:
```diff
@@ -0,0 +1,138 @@
+
+        Path dirPath = pathOfFileToWatch.getParent();
+        log.debug("dirPath is [{}]", dirPath);
+        if (dirPath == null) {
```
How can this be?

ravisharda replied:
According to [Path api docs](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Path.html), 
> Returns the parent path, or null if this path does not have a parent.

andreipaduroiu commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationPollingMonitor.java:
```diff
@@ -0,0 +1,138 @@
+    public void stopMonitoring() {
+        if (monitor != null) {
+            try {
+                monitor.stop(5 * 1000);
```
This method will interrupt the current thread if the timeout expires. This will happen regardless of whether you catch it or not. As a result, whomever else is running on this thread will not be able to execute subsequent code.

ravisharda replied:
To test your assertion, I temporarily reduced the timeout to 1ms, and then ran a sample shown below that invoked stopMonitoring. I didn't observe the behavior you are mentioning: `exiting`. Am I missing something?

```
 @Test
    public void test()  {
        try {
            Path dir = Files.createTempDirectory("fw-");
            File file = File.createTempFile("tf-", ".temp", dir.toFile());

            AtomicBoolean isCallbackInvoked = new AtomicBoolean(false);
            FileModificationMonitor watcher = new FileModificationPollingMonitor(file.toPath().toString(),
                    c -> isCallbackInvoked.set(true), 100);
            watcher.startMonitoring();
            Thread.sleep(2 * 1000);
            watcher.stopMonitoring();
            System.out.println("exiting");
        } catch(Exception e) {
            e.printStackTrace();
        }
```

andreipaduroiu commented on common/src/test/java/io/pravega/common/io/filesystem/FileModificationEventWatcherTests.java:
```diff
@@ -0,0 +1,58 @@
+
+        // Modify the watched file.
+        FileUtils.writeStringToFile(file, "hello", StandardCharsets.UTF_8, true);
+        watcher.join(5 * 1000); // Wait for max 5 seconds for the thread to die.
```
I suggest you make this a bit higher. You're likely to have failures on Travis (or our internal build) for **anything** that is time sensitive as those machines are shared, singled-core and thus anything can take unpredictably long. I would also suggest you make the base timeout configurable somehow.

ravisharda replied:
I've now modified it to 15 seconds.

andreipaduroiu commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaConnectionListener.java:
```diff
@@ -123,17 +151,14 @@ public PravegaConnectionListener(boolean ssl, String host, int port, StreamSegme
-                throw new RuntimeException(e);
-            }
+
+        final AtomicReference<SslContext> sslCtx;
```
This need not be AtomicReference. You are only setting its value once so there's no need to wrap it in order to access it in a lambda below.

ravisharda replied:
The SSL Context gets reloaded via the callback invoked by file modification event watcher or polling monitor. Depending on how the certificate is modified, there may be multiple occurrences of SSL context reloads.

andreipaduroiu commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaConnectionListener.java:
```diff
@@ -173,19 +205,73 @@ public void initChannel(SocketChannel ch) throws Exception {
          });
 
+        if (enableTls && enableTlsReload) {
+            log.info("TLS reload is enabled, so setting up a FileModificationMonitor object to monitor/watch changes in file: [{}]",
```
Too verbose of a message. Make it concise.

ravisharda replied:
Done.

andreipaduroiu commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaConnectionListener.java:
```diff
@@ -173,19 +205,73 @@ public void initChannel(SocketChannel ch) throws Exception {
+            tlsCertFileModificationMonitor = prepareCertificateMonitor(this.pathToTlsCertFile, this.pathToTlsKeyFile,
+                    sslCtx);
+            tlsCertFileModificationMonitor.startMonitoring();
+            log.info("Started the FileModificationMonitor object successfully");
```
"object"?

ravisharda replied:
Fixed

andreipaduroiu commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaConnectionListener.java:
```diff
@@ -173,19 +205,73 @@ public void initChannel(SocketChannel ch) throws Exception {
+                                                      AtomicReference<SslContext> sslCtx) {
+        return prepareCertificateMonitor(Files.isSymbolicLink(Paths.get(tlsCertificatePath)),
+                tlsCertificatePath, tlsKeyPath, sslCtx);
+
```
empty line

ravisharda replied:
Fixed.

andreipaduroiu commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaConnectionListener.java:
```diff
@@ -173,19 +205,73 @@ public void initChannel(SocketChannel ch) throws Exception {
+        FileModificationMonitor result;
+        try {
+            if (isTLSCertPathSymLink) {
+                log.info("The path to certificate file [{}] was found to be pointing to a symbolic link, " +
```
Too verbose.
Unnecessary comma after "so".

ravisharda replied:
Fixed.

andreipaduroiu commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaConnectionListener.java:
```diff
@@ -173,19 +205,73 @@ public void initChannel(SocketChannel ch) throws Exception {
+    @VisibleForTesting
+    FileModificationMonitor prepareCertificateMonitor(boolean isTLSCertPathSymLink, String tlsCertificatePath,
+                                                      String tlsKeyPath,
+                                                      AtomicReference<SslContext> sslCtx) {
```
Why is this AtomicReference? You do not need this.

ravisharda replied:
The SSL Context gets reloaded via the callback invoked by file modification event watcher or polling monitor. The SSL context might be in use while its reload gets triggered by the callback.

andreipaduroiu commented on segmentstore/server/src/main/java/io/pravega/segmentstore/server/security/TLSConfigChangeHandler.java:
```diff
@@ -0,0 +1,42 @@
+        sslContext.set(TLSHelper.newServerSslContext(pathToCertificateFile, pathToKeyFile));
+    }
+
+    AtomicInteger getNumOfConfigChangesSinceStart() {
```
Do not expose your internals to the outside. Provide a proper getter and setter (for `int`) to access this.

ravisharda replied:
Yeah, you are right. I've fixed it.

andreipaduroiu commented on segmentstore/server/src/main/java/io/pravega/segmentstore/server/security/TLSHelper.java:
```diff
@@ -0,0 +1,96 @@
+        Preconditions.checkNotNull(certificateFile);
+        Preconditions.checkNotNull(serverKeyFile);
+
+        Retry.RetryAndThrowExceptionally<IllegalStateException, Exception> retryConfig =
```
This isn't used.

ravisharda replied:
I've removed it now.

andreipaduroiu commented on segmentstore/server/src/main/java/io/pravega/segmentstore/server/security/TLSHelper.java:
```diff
@@ -0,0 +1,96 @@
+
+    private static void ensureExistAndAreReadable(File certificateFile, File serverKeyFile) {
+        if (!certificateFile.exists()) {
+            throw new IllegalStateException(String.format("Certificate file %s doesn't exist",
```
These look a lot more like `IllegalArgumentExceptions` since you are validating your arguments, not the internal state of your object.

I would also use `Preconditions.checkArgument` to make your code more compact.

ravisharda replied:
Yes, that makes sense. I've modified the code accordingly.

andreipaduroiu commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java:
```diff
@@ -0,0 +1,225 @@
+     *
+     * @param fileToWatch path of the file to watch
+     * @param callback          the callback to invoke when a modification to the {@code fileToWatch} is detected
+     * @param loopContinuously  whether to keep the thread look for file modification after one iteration. This option
```
"thread look"? Is it "thread lock"? If so please define this term as it's not standard.

ravisharda replied:
That was a typo. I've modified the comment now.

andreipaduroiu commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java:
```diff
@@ -0,0 +1,225 @@
+ *
+ */
+@Slf4j
+public class FileModificationEventWatcher extends Thread implements FileModificationMonitor {
```
Since you have chosen to use an entire thread for this task, can you please give it an appropriate name so that when we do a thread dump we know which thread is which?

ravisharda replied:
The thread already has a name given to it here: https://github.com/pravega/pravega/blob/f96e9ccfdf2c4faeeb06d289cefdbeeb888c837e/common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java#L112

andreipaduroiu commented:
@ravisharda I understand what this PR wants to achieve, however I am not comfortable with starting a new thread for each file watcher we want to create. Is there no API in Java (or other open-source library) that can do this asynchronously?

tkaitchuck commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java:
```diff
@@ -0,0 +1,238 @@
+                // See https://stackoverflow.com/questions/16777869/java-7-watchservice-ignoring-multiple-occurrences-
+                // of-the-same-event for a discussion on this topic.
+                //
+                // If the two events are not raised within this duration, the callback will be invoked twice, which we
```
The structure of this loop could be simplified a bit and prevent the issue which would avoid the confusion Fpj had below. If there were are private method factored out, this comment could be made into a javadoc.

ravisharda replied:
Sure, I've changed this part of the code. I first thought I should just take the comments and the `sleep` code out as you suggested, but then I was finding it hard to assign it a name that revealed its responsibility. So, I ended up extracting the watch key retrieval logic out too. The side effect, though, is that I couldn't take the comments out to method JavaDocs. Hope it looks OK.

ravisharda commented:
> @ravisharda I understand what this PR wants to achieve, however I am not comfortable with starting a new thread for each file watcher we want to create. Is there no API in Java (or other open-source library) that can do this asynchronously?

@Andrei I agree with you, it might be too expensive to create a thread-per-file, especially if we monitor modifications to several files. As of now, it is used only for monitoring a single file - the TLS certificate. I think we could live with it until we start watching multiple files, and refactor it when we hit an inflection point. I can create a separate enhancement issue if that works.

ravisharda commented on common/src/test/java/io/pravega/common/io/filesystem/FileModificationMonitorTests.java:
```diff
@@ -0,0 +1,89 @@
+
+import static io.pravega.test.common.AssertExtensions.assertThrows;
+
+public abstract class FileModificationMonitorTests {
```
@tkaitchuck `FileModificationEventWatcherTests` and `FileModificationPollingMonitorTests` test classes extend the abstract class `FileModificationMonitorTests`. `FileModificationMonitorTests` contains tests that are common across those two subclasses. 

Hope that clarifies.

ravisharda commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java:
```diff
@@ -0,0 +1,238 @@
+            }
+        } catch (InterruptedException | IOException e) {
+            logException(e);
+            throw new RuntimeException(e);
```
@tkaitchuck Thanks for spotting that error. I've fixed it now.

fpj commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java:
```diff
@@ -0,0 +1,249 @@
+ *   at times. If the calling code needs to be notified exactly once at all times, it should consider using
+ *   {@link FileModificationPollingMonitor} instead.
+ *
+ * - This class won't work for files that is under a symbolic link (pointing to a directory). This is because the
```
`This class won't work for a symbolic link...`

ravisharda replied:
Done.

fpj commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java:
```diff
@@ -0,0 +1,249 @@
+            assert directoryPath != null;
+            directoryPath.register(watchService, StandardWatchEventKinds.ENTRY_MODIFY,
+                    StandardWatchEventKinds.ENTRY_CREATE);
+            log.debug("Done setting up watch for modify entries for file at path: {}", this.watchedFilePath);
```
The message in the debug log message does not read well, please rephrase as I'm not what it is trying to say other than that it has set up the watch.

ravisharda replied:
I've fixed it now.

fpj commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java:
```diff
@@ -0,0 +1,249 @@
+                        break;
+                    }
+                } else {
+                    log.debug("watchKey for file at path {} was null", this.watchedFilePath);
```
If retrieveWatchKeyFrom is interrupted, then the catch block will allow the iteration to proceed and execute this log statement, which might not be meaningful in the case of an interruption. I'm essentially wondering whether we should be skipping the iteration altogether in the case of an interrupt. Maybe it is harmless, though.

ravisharda replied:
Sure, I have removed the log statement.

fpj commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java:
```diff
@@ -0,0 +1,249 @@
+        WatchKey result = watchService.take();
+        log.info("Retrieved and removed watch key for watching file at path: {}", this.watchedFilePath);
+
+        // Each file modification/create usually results in the WatcherService reporting the WatchEvent twice,
```
Do we know what kind of modification it has been?  We only care about content modification, we don't care about time stamp updates, and if we can differentiate the updates, then that's a better approach than adding a sleep. In the case it needs to be a sleep call, it is not clear why it needs to be 200ms. We don't expect frequent changes, so maybe it can be longer, although it clearly affects creation time.

ravisharda replied:
The 200 ms is a rather arbitrary value. A few examples I have seen used 50 or 100 ms. I kept 200 ms. just to make it more likely that the events will get deduplicated. I think 200 ms. is OK, based on all the testing I have done so far on Kubernetes clusters with Cert Manager.

fpj replied:
You've only answered the question about the sleep time, but the first part of the comment is more important for me because if that's possible, then we can eliminate this sleep altogether.

ravisharda replied:
Sure. Each file modification/create usually results in the WatcherService reporting the WatchEvent twice, as the file is updated twice: once for the content and once for the file modification time.  These events occur in quick succession. We should be able to differentiate the updates AFAIK, but then we'd need to keep track of the events in a stateful manner, which would make the code more complex. In most cases, the current deduplication logic should have the desired consequences; In the rare case where the two events are tackled nonetheless, there is not much downside except for creating an extra SSL Context and assigning it to the shared variable. Therefore, I think the current approach is a reasonable choice. 

Hope that answers your question.

fpj replied:
> We should be able to differentiate the updates AFAIK

If you are able to differentiate the type of change, then it is as simple as discarding whatever is not a content modification. In your argument, you talk about (1) content change and (2) timestamp change. If we are able to differentiate between (1) and (2), then we can simply discard (2) and only process (1). Why can't we do it this way and eliminate the sleep? This is sleep is causing unnecessary confusion in the code.

ravisharda replied:
@fpj I had misunderstood your question earlier. I meant to say previously that the two events - one for the content change and one for the timestamp update - are two separate events. We only get a `WatchEvent` that has a method `context()` which gives us the "relative path between the directory registered with the watch service, and the entry that is created, deleted, or modified"  ([src](https://docs.oracle.com/javase/8/docs/api/java/nio/file/WatchEvent.html#context--)). So, we can't distinguish between the two events except by indirect interpretation based on the time that they have arrived (stateful processing). Hope that clarifies.

EronWright replied:
At the end of the day, a hash of the content would suffice to distinguish a substantive update.  That's what the istio reference code that I shared does.

ravisharda replied:
I think what we have is sufficient for now. It works and does that without having to read the file based on JVM generated events alone. For cert manager distributed certificates we use the polling monitor instead of this class. I suggest we enhance this class later if needed.

fpj commented on segmentstore/server/src/main/java/io/pravega/segmentstore/server/security/TLSConfigChangeHandler.java:
```diff
@@ -0,0 +1,43 @@
+        sslContext.set(TLSHelper.newServerSslContext(pathToCertificateFile, pathToKeyFile));
+    }
+
+    int getNumOfConfigChangesSinceStart() {
```
Should this be an exposed metric?

ravisharda replied:
I don't think it'll be useful as a metric. It's just a convenience method for aid in unit testing as of now. Also, the metric is logged, to make it easier to identify when the SSL context is renewed.

fpj commented on segmentstore/server/src/main/java/io/pravega/segmentstore/server/security/TLSConfigChangeHandler.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    int getNumOfConfigChangesSinceStart() {
+
```
Please remove unnecessary new line.

ravisharda replied:
Done.

fpj commented on segmentstore/server/src/main/java/io/pravega/segmentstore/server/security/TLSConfigChangeHandler.java:
```diff
@@ -0,0 +1,43 @@
+    private @NonNull final String pathToCertificateFile;
+    private @NonNull final String pathToKeyFile;
+
+    public void handleTlsConfigChange() {
```
Isn't it sufficient that this is package protected rather than public?

ravisharda replied:
Yes, that's right. I've modified its access modifier, as suggested.

fpj commented on segmentstore/server/src/main/java/io/pravega/segmentstore/server/security/TLSConfigChangeHandler.java:
```diff
@@ -0,0 +1,43 @@
+
+    public void handleTlsConfigChange() {
+        log.info("Current reload count = {}", numOfConfigChangesSinceStart.incrementAndGet());
+        sslContext.set(TLSHelper.newServerSslContext(pathToCertificateFile, pathToKeyFile));
```
I'm not sure I understand this flow. This line sets a new context through `TLSHelper`, but I don't understand how this is actually loaded. Setting the atomic reference shouldn't cause it to be automatically loaded by Netty.

I'm also missing where the channels are being closed upon reload (that's what the description says).  I'm particularly concerned about it because dropping all connections at once will cause all clients to try to reconnect together, causing a herd effect for a large number of clients. I'm wondering if instead of closing all channels, we can:

- Not close them at all and simply create new channels with the new context
- Close channels gradually

ravisharda replied:
Sorry, the description still had a statement from the approach followed initially, which is now replaced with a different approach (based on the discussions we have had above): 

We now share the same SSL Context across all channels (like we did earlier). Upon detection that the certificate has changed, the callback creates a new SSL Context and assigns it to the shared variable. So, existing connections established using old SSL Context continue to function, while new ones use the new SSL Context built from the new certificate (and the key).

ravisharda commented on common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java:
```diff
@@ -0,0 +1,249 @@
+ *
+ * - If there are multiple file modifications in quick succession, only the last one may trigger the specified action.
+ *
+ * - The callback twice under rare circumstances when a file is modified. The Java {@link WatchService} raises two
```
Done.

ravisharda commented:
> Could you elaborate on what kind of testing has been performed in addition to unit tests for this PR?

In addition to unit tests, I've executed the following types of tests to verify the functionality:
1. Standalone mode cluster with client application writing to the cluster continuously. I replaced the certificate midway and check the client continues to function. Also, I run a new client application against the cluster. 
2. Docker-Compose clusters with the same scenario as above.
3. Kubernetes cluster with Cert Manager distributing certificates, with the same scenario as above.
